nova_vm/src/ecmascript/abstract_operations/operations_on_objects.rs

```diff
@@ -766,6 +766,45 @@ pub(crate) fn get_function_realm(
     }
 }
 
+/// ### [7.3.25 CopyDataProperties ( target, source, excludedItems )](https://tc39.es/ecma262/#sec-copydataproperties)
+/// The abstract operation CopyDataProperties takes arguments target (an Object), source (an
+/// ECMAScript language value), and excludedItems (a List of property keys) and returns either a
+/// normal completion containing unused or a throw completion.
+///
+/// NOTE: This implementation of CopyDataProperties takes an existing target object and populates
+/// it, but it does not support excluded items. It can be used to implement the spread operator in
+/// object literals, but not the rest operator in object destructuring.
+pub(crate) fn copy_data_properties(
+    agent: &mut Agent,
+    target: Object,
+    source: Value,
+) -> JsResult<()> {
+    // 1. If source is either undefined or null, return unused.
+    if source.is_undefined() || source.is_null() {
+        return Ok(());
+    }
+    // 2. Let from be ! ToObject(source).
+    let from = to_object(agent, source).unwrap();
+
+    // 3. Let keys be ? from.[[OwnPropertyKeys]]().
+    // 4. For each element nextKey of keys, do
+    for next_key in from.internal_own_property_keys(agent)? {
```

Comment:
**suggestion:** Reserve room in the object for these items using the `reserve` method on the ElementVectors.

Reply:
This would mean changing the type of `target` to be `OrdinaryObject`. Will do that in a follow-up PR if you don't mind.